File: Zeze/Transaction/Collections/PMap2.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Diagnostics;

namespace Zeze.Transaction.Collections
{
    public sealed class PMap2<K, V> : PMap<K, V> where V : Bean
    {
        public PMap2(long logKey, Func<ImmutableDictionary<K, V>, Log> logFactory) : base(logKey, logFactory)
        {
        }

        public override V this[K key]
        {
            get => Data[key];
            set
            {
                if (value == null)
                    throw new ArgumentNullException();

                if (this.IsManaged)
                {
                    value.InitTableKey(TableKey, Parent);
                    value.VariableId = this.VariableId;
                    var txn = Transaction.Current;
                    var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : map;
                    var newv = oldv.SetItem(key, value);
                    if (newv != oldv)
                    {
                        txn.PutLog(NewLog(newv));
                        ((ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteMap2<K, V>(this))).LogPut(key, value);
                    }
                }
                else
                {
                    map = map.SetItem(key, value);
                }

            }
        }

        public override void Add(K key, V value)
        {
            if (key == null)
                throw new ArgumentNullException();
            if (value == null)
                throw new ArgumentNullException();

            if (this.IsManaged)
            {
                value.InitTableKey(TableKey, Parent);
                value.VariableId = this.VariableId;
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : map;
                var newv = oldv.Add(key, value);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ((ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteMap2<K, V>(this))).LogPut(key, value);
                }
            }
            else
            {
                map = map.Add(key, value);
            }
        }

        public override void AddRange(IEnumerable<KeyValuePair<K, V>> pairs)
        {
            foreach (KeyValuePair<K, V> p in pairs)
            {
                if (p.Key == null)
                    throw new ArgumentNullException();
                if (p.Value == null)
                    throw new ArgumentNullException();
            }

            if (this.IsManaged)
            {
                foreach (var p in pairs)
                {
                    p.Value.InitTableKey(TableKey, Parent);
                    p.Value.VariableId = this.VariableId;
                }
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : map;
                var newv = oldv.AddRange(pairs);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ChangeNoteMap2<K, V> note = (ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteMap2<K, V>(this));
                    foreach (var p in pairs)
                    {
                        note.LogPut(p.Key, p.Value);
                    }
                }
            }
            else
            {
                map = map.AddRange(pairs);
            }
        }

        public override void SetItem(K key, V value)
        {
            if (key == null)
                throw new ArgumentNullException();
            if (value == null)
                throw new ArgumentNullException();

            if (this.IsManaged)
            {
                value.InitTableKey(TableKey, Parent);
                value.VariableId = this.VariableId;
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : map;
                var newv = oldv.SetItem(key, value);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ((ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteMap2<K, V>(this))).LogPut(key, value);
                }
            }
            else
            {
                map = map.SetItem(key, value);
            }
        }

        public override void SetItems(IEnumerable<KeyValuePair<K, V>> pairs)
        {
            foreach (KeyValuePair<K, V> p in pairs)
            {
                if (p.Key == null)
                    throw new ArgumentNullException();
                if (p.Value == null)
                    throw new ArgumentNullException();
            }

            if (this.IsManaged)
            {
                foreach (var p in pairs)
                {
                    p.Value.InitTableKey(TableKey, Parent);
                    p.Value.VariableId = this.VariableId;
                }
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : map;
                var newv = oldv.SetItems(pairs);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ChangeNoteMap2<K, V> note = (ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteMap2<K, V>(this));
                    foreach (var p in pairs)
                    {
                        note.LogPut(p.Key, p.Value);
                    }
                }
            }
            else
            {
                map = map.SetItems(pairs);
            }
        }

        public override void Add(KeyValuePair<K, V> item)
        {
            if (item.Key == null)
                throw new ArgumentNullException();
            if (item.Value == null)
                throw new ArgumentNullException();

            if (this.IsManaged)
            {
                item.Value.InitTableKey(TableKey, Parent);
                item.Value.VariableId = this.VariableId;
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : map;
                var newv = oldv.Add(item.Key, item.Value);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ((ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteMap2<K, V>(this))).LogPut(item.Key, item.Value);
                }
            }
            else
            {
                map = map.Add(item.Key, item.Value);
            }
        }

        public override void Clear()
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : map;
                if (!oldv.IsEmpty)
                {
                    ChangeNoteMap2<K, V> note = (ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteMap2<K, V>(this));
                    foreach (var e in oldv)
                    {
                        note.LogRemove(e.Key);
                    }
                    txn.PutLog(NewLog(ImmutableDictionary<K, V>.Empty));
                }
            }
            else
            {
                map = ImmutableDictionary<K, V>.Empty;
            }
        }

        public override bool Remove(K key)
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : map;
                var newv = oldv.Remove(key);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ((ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteMap2<K, V>(this))).LogRemove(key);
                    return true;
                }
                else
                {
                    return false;
                }
            }
            else
            {
                var old = map;
                map = map.Remove(key);
                return old != map;
            }
        }

        public override bool Remove(KeyValuePair<K, V> item)
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : map;
                if (oldv.TryGetValue(item.Key, out var olde) && olde.Equals(item.Value))
                {
                    var newv = oldv.Remove(item.Key);
                    txn.PutLog(NewLog(newv));
                    ((ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteMap2<K, V>(this))).LogRemove(item.Key);
                    return true;
                }
                else
                {
                    return false;
                }
            }
            else
            {
                if (map.TryGetValue(item.Key, out var oldv) && oldv.Equals(item.Value))
                {
                    map = map.Remove(item.Key);
                    return true;
                }
                else
                {
                    return false;
                }
            }
        }

        protected override void InitChildrenTableKey(TableKey tableKey)
        {
            foreach (var v in map.Values)
            {
                v.InitTableKey(TableKey, Parent);
            }
        }
    }
}

File: ZezeJava/src/Zeze/Transaction/Collections/PMap2.java
package Zeze.Transaction.Collections;

import Zeze.Transaction.*;
import java.util.*;
import org.pcollections.Empty;

public final class PMap2<K, V extends Bean> extends PMap<K, V> {
	public PMap2(long logKey, LogFactory<org.pcollections.PMap<K, V>> logFactory) {
		super(logKey, logFactory);
	}

	@Override
	public void putAll(Map<? extends K, ? extends V> m) {
		for (var p : m.entrySet()) {
			if (p.getKey() == null) {
				throw new NullPointerException();
			}
			if (p.getValue() == null) {
				throw new NullPointerException();
			}
		}

		if (this.isManaged()) {
			for (var p : m.entrySet()) {
				p.getValue().InitRootInfo(RootInfo, getParent());
				p.getValue().setVariableId(this.getVariableId());
			}
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldm = null != log ? ((LogV)log).Value : map;
			var newm = oldm.plusAll(m);
			if (newm != oldm) {
				txn.PutLog(NewLog(newm));
				@SuppressWarnings("unchecked")
				ChangeNoteMap2<K, V> note = (ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteMap2<K, V>(this));
				for (var p : m.entrySet()) {
					note.LogPut(p.getKey(), p.getValue());
				}
			}
		}
		else {
			map = map.plusAll(m);
		}
	}

	@SuppressWarnings("unchecked")
	@Override
	public V put(K key, V value) {
		if (key == null) {
			throw new NullPointerException();
		}
		if (value == null) {
			throw new NullPointerException();
		}

		if (this.isManaged()) {
			value.InitRootInfo(RootInfo, getParent());
			value.setVariableId(this.getVariableId());
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			var oldm = null != log ? ((LogV)log).Value : map;
			var oldv = oldm.get(key);
			var newm = oldm.plus(key, value);
			if (newm != oldm) {
				txn.PutLog(NewLog(newm));
				((ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteMap2<K, V>(this))).LogPut(key, value);
			}
			return oldv;
		}
		else {
			var oldv = map.get(key);
			map = map.plus(key, value);
			return oldv;
		}
	}

	@Override
	public void clear() {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldm = null != log ? ((LogV)log).Value : map;
			if (!oldm.isEmpty()) {
				@SuppressWarnings("unchecked")
				ChangeNoteMap2<K, V> note = (ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteMap2<K, V>(this));
				for (var e : oldm.entrySet()) {
					note.LogRemove(e.getKey());
				}
				txn.PutLog(NewLog(Empty.map()));
			}
		}
		else {
			map = Empty.map();
		}
	}

	@SuppressWarnings("unchecked")
	@Override
	public boolean remove(K key) {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			var oldm = null != log ? ((LogV)log).Value : map;
			var newm = oldm.minus(key);
			if (newm != oldm) {
				txn.PutLog(NewLog(newm));
				((ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteMap2<K, V>(this))).LogRemove(key);
				return true;
			}
			else {
				return false;
			}
		}
		else {
			var old = map;
			map = map.minus(key);
			return old != map;
		}
	}

	@SuppressWarnings("unchecked")
	@Override
	public boolean remove(Map.Entry<K, V> item) {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			var oldm = null != log ? ((LogV)log).Value : map;
			Object olde = oldm.get(item.getKey());
			if (null == olde)
				return false;
			if (olde.equals(item.getValue())) {
				var newm = oldm.minus(item.getKey());
				txn.PutLog(NewLog(newm));
				((ChangeNoteMap2<K, V>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteMap2<K, V>(this))).LogRemove(item.getKey());
				return true;
			}
			else {
				return false;
			}
		}
		else {
			Object oldv = map.get(item.getKey());
			if (oldv == null)
				return false;
			if (oldv.equals(item.getValue())) {
				map = map.minus(item.getKey());
				return true;
			}
			else {
				return false;
			}
		}
	}

	@Override
	protected void InitChildrenRootInfo(Zeze.Transaction.Record.RootInfo tableKey) {
		for (var v : map.values()) {
			v.InitRootInfo(RootInfo, getParent());
		}
	}
}